### PR TITLE
ExPlat: Add /lib anon-id

### PR DESCRIPTION
**Pull request by jessie-ross**
#### Changes proposed in this Pull Request

This PR adds `/lib/explat` anon-id functionality. 

Part of a series rebuilding the reference WIP PR #48475 (Although no longer kept in sync).

#### Testing instructions

* Added unit tests

### CODE REVIEW

File: client/lib/explat/internals/anon-id.ts
Comment (jessie-ross) [inline client/lib/explat/internals/anon-id.ts:32]: Is this timing ok?

File: client/lib/explat/internals/anon-id.ts
Comment (yanirs) [inline client/lib/explat/internals/anon-id.ts]: Related to https://github.com/Automattic/wp-calypso/pull/50261#discussion_r580644097: If the only reason we're exporting `isDevelopmentMode` from `log-error.ts` is for this function, then it'd be better to move this function there, as it's a more natural fit. In any case, it's strange that `isDevelopmentMode` is a part of the interface of `log-error.ts`.

File: client/lib/explat/internals/anon-id.ts
Comment (yanirs) [inline client/lib/explat/internals/anon-id.ts]: Why are we using these functions rather than [the functions from the `@automattic/calypso-analytics` package](https://github.com/Automattic/wp-calypso/tree/trunk/packages/calypso-analytics)?

File: client/lib/explat/internals/anon-id.ts
Comment (yanirs) [inline client/lib/explat/internals/anon-id.ts]: Yeah, I'm not sure that this code is required in its current form, and if it is, it shouldn't be an ExPlat-only thing. 

Looking at the `calypso-analytics` package, it exposes [`initializeAnalytics()`](https://github.com/Automattic/wp-calypso/blob/da5b69a6ed89e7494bf468b0ff25ff0763e49376/packages/calypso-analytics/src/tracks.ts#L131), which seems to guarantee that the `tk_ai` cookie is set if needed (either by the server or by a client-side fallback).

I'm not sure if it's safe to call `initializeAnalytics()` multiple times. If it isn't, we can fix it (or add a wrapper) to the `calypso-analytics` package and call that, obviating the need for this function.

**Review by yanirs — CHANGES_REQUESTED** [review]
As noted inline, there should be a better way to verify that Tracks has been loaded (or blocked) and that the `tk_ai` cookie is set to its expected value. Let's rely on existing packages for that instead of the polling solutions.

File: client/lib/explat/internals/anon-id.ts
Comment (jessie-ross) [reply client/lib/explat/internals/anon-id.ts]: Very good point! I didn't look deeply enough into the tracks code and should have at least asked about this 🤦

File: client/lib/explat/internals/anon-id.ts
Comment (jessie-ross) [reply client/lib/explat/internals/anon-id.ts]: Fair enough for `tracksAnonymousUserId` but for `recordTracksEvents` it makes sense to use it as it is the de-facto way of sending tracks events in Calypso and it seems to add functionality on.

File: client/lib/explat/internals/anon-id.ts
Comment (jessie-ross) [reply client/lib/explat/internals/anon-id.ts]: It seems like `initializeAnalytics()` sets up tracks but it does not initialise the `anonId`, for that it needs to either send an initial tracks event - which it doesn't do unless there is given user data AFAICT, or send an `identifyAnonUser` tracks command.
 
What might be misleading is that it returns a promise that comes from [`checkForBlockedTracks()`](https://github.com/Automattic/wp-calypso/blob/3c6f806059255c46cdfe58bba841dfb652690302/packages/calypso-analytics/src/tracks.ts#L78-L109), this doesn't check that tracks events have been sent but rather whether `w2.js` has loaded properly - e.g. blocked by an ad-blocker as `w2.js` isn't yet bundled with Calypso.

**Comment by jessie-ross** [comment]
> As noted inline, there should be a better way to verify that Tracks has been loaded (or blocked) and that the tk_ai cookie is set to its expected value. Let's rely on existing packages for that instead of the polling solutions.

The core of tracks is `w2.js`. AFAICT the closest thing that is exported to what we need is `identifyAnonUser` which we can call like so `window.wpcom.tracks.identifyAnonUser()`, this however won't generate a new anonId if one doesn't exist.

I might mention @bperson in here as he seems to be the maintainer: We are looking for a way to await and get the tracks `anonId` from within Calypso, possibly at boot time, initialising it if necessary. Is there a way to do so that is preferential to polling for it?

**Comment by bperson** [comment]
> I might mention @bperson in here as he seems to be the maintainer: We are looking for a way to await and get the tracks anonId from within Calypso, possibly at boot time, initialising it if necessary. Is there a way to do so that is preferential to polling for it?

No and the whole setup of Tracks in client code is to get out of the way. If `w2.js` is loaded last, while we accumulate commands in `window._tkq`, that is completely fine and expected. Client code shouldn't care whether an analytics event was successfully triggered ( you shouldn't ask the user to re-trigger an action because an analytics call failed, or make him wait for analytics code to load before doing an action ) and you will not get a `tk_ai` cookie created until `w2.js` is loaded.

I seem to remember a similar discussion a couple of years ago though, so I'll search today if there isn't any existing art for this but imo it's messing with one of the fundamental assumptions on which Tracks was built ... and thus relying on `tk_ai` might not be the right move.

**Comment by jessie-ross** [comment]
I think polling is a reasonable path forward at this point in time, the only concern I have about it is performance, but then again it is fairly small taking very little time per poll and polling relatively infrequently.

**Comment by jessie-ross** [comment]
Hmm, alternatively we could generate our own `anonId` and send our own alias event.

**Comment by withinboredom** [comment]
I think the best solution would be to generate our own anon id, then append it in the `aip` cookie. Then when the user logs in, tracks will capture that anon id.

File: client/lib/explat/internals/anon-id.ts
Comment (yanirs) [reply client/lib/explat/internals/anon-id.ts]: I'm not sure about that given [this deprecation note](https://github.com/Automattic/wp-calypso/blob/61acc957638fc7009632af4c0c058aa6bc0e82b7/packages/calypso-analytics/README.md#deprecates). It's worth asking around where things stand.

**Comment by yanirs** [comment]
@jessie-ross can you please summarise and move this discussion to a P2 thread?

Adding thoughts here for now.

> I think polling is a reasonable path forward at this point in time, the only concern I have about it is performance, but then again it is fairly small taking very little time per poll and polling relatively infrequently.

Won't the client be blocked until we get the ID, which can be five seconds until we can query and render the experiment?

> I think the best solution would be to generate our own anon id, then append it in the `aip` cookie. Then when the user logs in, tracks will capture that anon id.

I'm not sure that generating an anon ID when one may already exist is the best approach, as assignment to experiments depends on the ID. However, if there isn't a `tk_ai` cookie, what's the harm in creating it ourselves?

In any case, if we do go down the path of cookie generation, I think that the code to do it should live in `calypso-analytics` rather than in our client.

**Comment by jessie-ross** [comment]
P2'd at pbmo2S-GA-p2

**Comment by jessie-ross** [comment]
As part of this implementation I should additionally stop polling if `w2.js` doesn't load.

File: packages/calypso-analytics/src/tracks.ts
Comment (jessie-ross) [inline packages/calypso-analytics/src/tracks.ts:116]: @scinos might get you to check this as I am not familiar with this package.

Comment (jessie-ross) [reply packages/calypso-analytics/src/tracks.ts:116]: Do I need to bump the package version? I am not familiar with how the package deps update.

Comment (scinos) [reply packages/calypso-analytics/src/tracks.ts:116]: No need to bump the package version if you just want to use this new functionality inside the monorepo.

We only need to bump the version if/when we are going to do a public release of the package. I can see the latest release was done by @sarayourfriend ~7 months ago, maybe she knows more about this package?

Comment (sarayourfriend) [reply packages/calypso-analytics/src/tracks.ts:116]: Sorry, I don't know anything about the package, I just released it as part of a mass release of packages that needed releasing.

File: client/lib/explat/internals/anon-id.ts
Comment (yanirs) [reply client/lib/explat/internals/anon-id.ts:32]: As discussed on the P2 thread, I think it's OK, but it's worth adding an explanation here as an inline comment because five seconds sounds like a long time in general.

File: client/lib/explat/internals/anon-id.ts
Comment (yanirs) [reply client/lib/explat/internals/anon-id.ts]: Alright, given that we decided to go with the polling, can you please remove the TODOs?

File: client/lib/explat/internals/test/anon-id.ts
Comment (yanirs) [inline client/lib/explat/internals/test/anon-id.ts]: Hmm... This may be a behaviour change from the current assumptions for experiments that cross the login boundary. However, I think it's covered by the assignment cookie, and in general we can't rely on the anon ID cookie being around forever. We should also be able to access the aliasing history on the backend if we move away from assignment cookies, so I think this is fine even if it's not the same as the existing client.

@withinboredom may have more thoughts.

File: client/lib/explat/internals/anon-id.ts
Comment (yanirs) [inline client/lib/explat/internals/anon-id.ts]: This comment can use a bit of tidying up, assuming that the reader has less context.

**Review by yanirs — APPROVED** [review]
I think it's good to go apart from a bit of tidying up of the comments. Should also sort out the automated check failures.

File: client/lib/explat/internals/test/anon-id.ts
Comment (jessie-ross) [reply client/lib/explat/internals/test/anon-id.ts]: I think it won't hurt to do a quick check for anonId before we check for logged-in.

**Comment by ciampo** [comment]
Hey @jessie-ross! Just a reminder that, since this PR included changes that affect ETK, it's best practice to test for ETK regressions and deploy a new ETK version after merging a PR (for more details, see p58i-a7j-p2)

**Comment by jessie-ross** [comment]
Hi @ciampo, not too sure how this change affects ETK? Could it be that I may have accidentally added commits that touched it during a bad merge and auto-added ETK team as reviewers?

**Comment by ciampo** [comment]
> not too sure how this change affects ETK? 

It may affect ETK _marginally_ because of [this change](https://github.com/Automattic/wp-calypso/pull/50280/files#diff-0889a4603afce1fe9bdc82ea139640ea0ff646e3fe2b52d4792b1c84d0012d7bR111-R117) to the `@automattic/calypso-analytics` package — which is a [dependency](https://github.com/Automattic/wp-calypso/blob/trunk/apps/editing-toolkit/package.json#L92) of ETK.

Since the only change the `@automattic/calypso-analytics` package was the addition of a new function, it likely didn't cause any changes to the ETK bundle — so excuse me if my message made it seem like a bigger deal.

I guess that, with the introduction of automated checks (as described in p58i-a7j-p2), it will be easier to know wether a merged PR actually affected the compiled bundle of ETK.

**Pull request merged**
